src/nano_contracts/builder.ts

```diff
@@ -324,47 +441,104 @@ class NanoContractTransactionBuilder {
-          `Method needs ${methodArgs.length} parameters but data has ${this.args.length}.`
-        );
-      }
+  /**
+   * Build a transaction object from the built inputs/outputs/tokens
+   *
+   * It will create a Transaction or CreateTokenTransaction, depending on the vertex type
+   *
+   * @throws {NanoContractTransactionError} In case the create token options is null
+   *
+   * @memberof NanoContractTransactionBuilder
+   * @inner
+   */
+  async buildTransaction(
+    inputs: IDataInput[],
+    outputs: IDataOutput[],
+    tokens: string[]
+  ): Promise<Transaction | CreateTokenTransaction> {
+    if (this.vertexType === NanoContractVertexType.TRANSACTION) {
+      return transactionUtils.createTransactionFromData(
+        {
+          version: DEFAULT_TX_VERSION,
+          inputs,
+          outputs,
+          tokens,
+        },
+        this.wallet.getNetworkObject()
+      );
+    }
 
-      for (const [index, arg] of methodArgs.entries()) {
-        const serialized = serializer.serializeFromType(this.args[index], arg.type);
-        serializedArgs.push(serialized);
-      }
+    if (this.createTokenOptions === null) {
+      throw new NanoContractTransactionError(
+        "Create token options can't be null when creating a create token transaction."
+      );
     }
 
+    // It's a token creation transaction
+    // then we get the token creation data from the utils method
+    // and concatenate the nano actions inputs/outputs/tokens
+    const data = await tokensUtils.prepareCreateTokenData(
+      this.createTokenOptions.mintAddress,
+      this.createTokenOptions.name,
+      this.createTokenOptions.symbol,
+      this.createTokenOptions.amount,
+      this.wallet.storage,
+      {
+        changeAddress: this.createTokenOptions.changeAddress,
+        createMint: this.createTokenOptions.createMint,
+        mintAuthorityAddress: this.createTokenOptions.mintAuthorityAddress,
+        createMelt: this.createTokenOptions.createMelt,
+        meltAuthorityAddress: this.createTokenOptions.meltAuthorityAddress,
+        data: this.createTokenOptions.data,
+        isCreateNFT: this.createTokenOptions.isCreateNFT,
+        skipDepositFee: this.createTokenOptions.contractPaysTokenDeposit,
+      }
+    );
+
+    data.inputs = concat(data.inputs, inputs);
+    data.outputs = concat(data.outputs, outputs);
+    data.tokens = uniq(concat(data.tokens, tokens));
+
+    return transactionUtils.createTransactionFromData(data, this.wallet.getNetworkObject());
+  }
+
+  /**
+   * Build a full transaction with nano headers from nano contract data
+   *
+   * @throws {NanoContractTransactionError} In case the arguments to build the tx are invalid
+   *
+   * @memberof NanoContractTransactionBuilder
+   * @inner
+   */
+  async build(): Promise<Transaction> {
+    await this.verify();
+
+    // Transform actions into inputs and outputs
+    const [inputs, outputs, tokens] = await this.buildInputsOutputs();
```

Comment:
Maybe this method could also build the `actions` array, so the `if (this.actions)` below would not be needed

Reply:
I can't do this because I need the full array of tokens to get the correct `tokenIndex`, which may change in the `buildTransaction` method.

Actually this an interesting test, having an action for a token that is not HTR and a new token being created in the same tx.